### ISSUE TITLE
XML to Json serialization is not mapping the data properly

### ISSUE THREAD

**Issue by Krishnamohan10** [FasterXML/jackson-dataformat-xml#229]
I have some list months in the row tags in a XML format. But while mapping the data to Json it is not mapping properly. Please check below for the reference.

```
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.xml.XmlMapper;

public class XML2Json {

	public static void main(String[] args) throws JsonProcessingException, IOException {

		String xml = "<root><months><row>Apr 2013</row><row>May 2013</row><row>Jun 2013</row><row>Jul 2013</row><row>Aug 2013</row></months></root>";

		XmlMapper xmlMapper = new XmlMapper();
		JsonNode node = xmlMapper.readTree(xml.getBytes());

		ObjectMapper jsonMapper = new ObjectMapper();
		String json = jsonMapper.writeValueAsString(node);

		System.out.println(json);
	}
}
```
The output for the above code is:
 ```
   "months": {
      "row": "Aug 2013"
    }
```

**Comment by capey0** [comment]
I get the same kind of thing from this?

```
String xmlX = "<top><hdr>head</hdr><recs><rec><fn>aa</fn><il>bb</il><ln>cc</ln></rec><rec><fn>dd</fn><il/><ln>ee</ln></rec><rec><fn>dd</fn><il></il><ln>ee</ln></rec></recs></top>";
JsonNode outJSON = null;
ObjectMapper om = new ObjectMapper();
XmlMapper xM = new XmlMapper();

try {
	System.out.println("Incoming XML: " + xmlX);
	outJSON = om.readTree(om.writeValueAsString(xM.readTree(xmlX.getBytes())));
	System.out.println("Outgoing JSON: " + om.writeValueAsString(outJSON));
}
catch (Exception e) {
}


=====================
Here is the output:
=====================
Incoming XML: <top><hdr>head</hdr><recs><rec><fn>aa</fn><il>bb</il><ln>cc</ln></rec><rec><fn>dd</fn><il/><ln>ee</ln></rec><rec><fn>dd</fn><il></il><ln>ee</ln></rec></recs></top>
Outgoing JSON: {"hdr":"head","recs":{"rec":{"fn":"dd","il":null,"ln":"ee"}}}

```

**Comment by absolutesantaja** [comment]
This issue still exists with 2.9.3. I'm trying to migrate from json.org to jackson and the output of the conversion is incomplete.

**Comment by cowtowncoder** [comment]
Code works exactly as expected; there will not be "fixes". I'll break this down.

1. Tree Model (`JsonNode`) is not fully supported by module: and in particular, there is no real way to reliably tell apart expected Array and Object from XML input: XML only has elements and attributes. Databind can infer distinction from Java type (array, `Collection`). Tree Model can only assume Object.
2. XML only allows single root, and this element represents Object container which in JSON would be pair of `{` and `}`.

From this I would strongly suggest that attempts to directly convert between XML and JSON using Tree Model is a bad idea. It is fine to convert from XML to POJO to JSON, or vice versa -- this will retain enough metadata to allow for proper conceptual conversion.
Fundamental problem here is not Jackson, or its handling, but rather that logical data model in XML is quite different from JSON and it simply can not (and should) be directly converted. It will not work.

**Comment by absolutesantaja** [comment]
The problem is the json.org library does the conversion successfully in the examples I'm working with but the license isn't usable on our project and all the stack overflow posts point to Jackson as the only alternative. XLST is an alternative but i've been trying to avoid going down that path.

**Comment by cowtowncoder** [comment]
@absolutesantaja one common case where things get wrong quickly is the case of "one element is Object; two or more repeated is Array" logic. This usually causes problems by converting what should be 1-value into Object/Map.

But as to converting from logic used by `json.org` it would be doable using Streaming API (creating `JsonParser` -- actually sub-class, `FromXmlParser` but usually base API sufficient).
That is some extra work but possibly worth the hassle. Parser would expose all elements, including ones that would be lost as duplicates. And then builder could create whatever data structure is needed, including `JsonNode` -- taking into account necessary coercion for duplicates, after establishing there is more than one instance of what seems like same property.

One thing that could be done for Jackson 3.0 -- and I'll make note of this as an idea -- would be to have separate XML Tree Model (`XmlNode`) which could implement same base interfaces (`TreeNode`, essentially), but allow duplicates, and also allow coercion into Arrays using common heuristics. I don't know if that should occur when reading, or as output/conversion option.

Since it seems there is a lot of demand for something like this I think it should be tackled in some form for Jackson 3.x. It is tricky challenge due to JSON/XML data model impedance... but worth tackling I think.

**Comment by Adiqq** [comment]
I'd like to add that 

> Code works exactly as expected; there will not be "fixes".

Does not solve existing problem. We have problem here that might be hard to solve due to technical problems, but take perspective of end-user using this library.
I just want to convert XML to JSON, with some constraints, I don't want creating any POJO, because in most cases I will not know structure of incoming XML. In this case there might be some configurable guess, how it should be mapped. Additionaly I can provide XSD file for well-defined mapping for specific case. I also want to keep root tag from XML in result JSON. I might care for edge cases later, once I can solve my initial problem.

Now, what solution do I have right now, without studying intrinsics of Jackson and this XML extension for it?

**Comment by Adiqq** [comment]
I prepared small Kotlin [library](https://github.com/Adiqq/xml-to-json-converter) to overcome problems mentioned in this issue. It offers naive implementation to overcome problem with missing root tag and it maps multiple child elements with the same name into a array.

**Comment by Cosmos19** [comment]
I land up with the same scenario today. It would be great if we have any **workaround** or **solution** for this usecase. I cant fall back from Jackson because the amount of work i put to use this high.

Thanks in advance.

**Comment by cowtowncoder** [comment]
As things are, I will close this as "work as intended" just because `JsonNode` use is not supported (and issue with repeated elements) is mentioned. This does not mean that I don't think there is room for improvement, but rather that there needs to be separate issue(s) for specifically improving behavior in certain way. I think some already exist. Ideas include:

1. Separate new Xml-specified tree/node model, not based on `JsonNode` (but might extend `TreeNode` if that matters)
2. Alternate handling mode in which repeated elements read are used to "convert" `ObjectNode` into `ArrayNode` (would probably  -- I will actually file `jackson-databind` issue for that.

**Comment by cowtowncoder** [comment]
For what it is worth, this ultimately became #403 which is (2) from above: it will be included in 2.12.0.
